hummingbot/strategy/execution4/execution4.pyx

```diff
@@ -0,0 +1,358 @@
+                )
+            else:
+                market_order_record = self._sb_order_tracker.c_get_market_order(market_info, order_id)
+                self.log_with_clock(
+                    logging.INFO,
+                    f"({market_info.trading_pair}) Market sell order {order_id} "
+                    f"({market_order_record.amount} {market_order_record.base_asset}) has been completely filled."
+                )
+
+    cdef c_start(self, Clock clock, double timestamp):
+        StrategyBase.c_start(self, clock, timestamp)
+        self.logger().info(f"Waiting for {self._time_delay} to place orders")
+        self._start_timestamp = timestamp
+        self._last_timestamp = timestamp
+
+    cdef c_tick(self, double timestamp):
+        StrategyBase.c_tick(self, timestamp)
+        cdef:
+            int64_t current_tick = <int64_t>(timestamp // self._status_report_interval)
+            int64_t last_tick = <int64_t>(self._last_timestamp // self._status_report_interval)
+            bint should_report_warnings = ((current_tick > last_tick) and
+                                           (self._logging_options & self.OPTION_LOG_STATUS_REPORT))
+            list active_maker_orders = self.active_maker_orders
+
+        try:
+            if not self._all_markets_ready:
+                self._all_markets_ready = all([market.ready for market in self._sb_markets])
+                if not self._all_markets_ready:
+                    # Markets not ready yet. Don't do anything.
+                    if should_report_warnings:
+                        self.logger().warning(f"Markets are not ready. No market making trades are permitted.")
+                    return
+
+            if should_report_warnings:
+                if not all([market.network_status is NetworkStatus.CONNECTED for market in self._sb_markets]):
+                    self.logger().warning(f"WARNING: Some markets are not connected or are down at the moment. Market "
+                                          f"making may be dangerous when markets or networks are unstable.")
+
+            for market_info in self._market_infos.values():
+                self.c_process_market(market_info)
+        finally:
+            self._last_timestamp = timestamp
+
+    cdef c_place_orders(self, object market_info):
+        cdef:
+            MarketBase market = market_info.market
+            curr_order_amount = min(self._order_amount / self._num_individual_orders, self._quantity_remaining)
+            object quantized_amount = market.c_quantize_order_amount(market_info.trading_pair, Decimal(curr_order_amount))
+            object quantized_price = market.c_quantize_order_price(market_info.trading_pair, Decimal(self._order_price))
+
+        self.logger().info(f"Checking to see if the incremental order size is possible")
+        self.logger().info(f"Checking to see if the user has enough balance to place orders")
+
+        if (quantized_amount != 0):
```

Comment:
Another coding style issue... we don't add parentheses around simple if / while clauses in Python like in C or Java. If it's a multi-line clause, then it's acceptable as a way to organize a very long if / while clause. But if it's short like this we don't do parentheses.